# MCIS6273 Data Mining (Prof. Maull) / Fall 2017 / HW1 SOLUTION

**This assignment is worth up to 20 POINTS to your grade total if you complete it on time.**

| Points <br/>Possible | Due Date | Time Commitment <br/>(estimated) |
|:---------------:|:--------:|:---------------:|
| 20 | Saturday, Sep 30 @ Midnight | _up to_ 10 hours |


* **GRADING:** Grading will be aligned with the completeness of the objectives.

* **INDEPENDENT WORK:** Copying, cheating, plagiarism  and academic dishonesty _are not tolerated_ by Univerisity or course policy.  Please see the syllabus for the full departmental and University statement on the academic code of honor.

## OBJECTIVES
* Work with core Pandas and Scikit-Learn concepts in data, data types, representation of data and plotting data

* Explore concepts in statistical inference over real data within Scikit-Learn

* Work with data to understand distance metrics in Scikit-Learn and the impact various metrics have on the outcomes

## WHAT TO TURN IN
You are being encouraged to turn the assignment in using the provided
Jupyter Notebook.  To do so, clone
[the course repository](https://github.com/kmsaumcis/mcis6273_f17_datamining) and modify
the `hw1.ipynb` file in the `homework/hw1` directory.  If you do not know
how to do this, please ask, or visit one of the many tutorials out there
on the basics of using Github and cloning repositories.

Turn in a copy of a `.ipynb` file, a PDF or Word Document to Blackboard
with the answers to the questions labeled with the &#167; sign.

## ASSIGNMENT TASKS
### (20%) Work with core Pandas and Scikit-Learn concepts in data, data types, representation of data and plotting data 

A great deal of time doing data mining involves understanding the data in a dataset and preprocessing it in preparation for working with it in a real analysis of some sort.  We will develop an understanding of:

  * how to empirically perform and understand the descriptive statistics of a dataset,
  * understand how to reason about data features,
  * understand how to use distance metrics,
  * understand binarizaration contexts abd techniques, and
  * understand how to randomly sample datasets to understand the distribution of data.

We'll be working a new  much smaller dataset for this task.  Please check the repository for the [`bank.csv` data set](https://github.com/kmsaumcis/mcis6273_f17_datamining/blob/master/homework/hw1/bank-data.csv).  There are fewer than 1000 rows to work with in this data.  You will need to use [Pandas](http://pandas.pydata.org/) and [Scikit-Learn](http://scikit-learn.org/stable/index.html) for all the questions here, and it is recommended you take a look at the optional cheat sheets [for lecture 2 in the syllabus](https://kmsaumcis.gitbub.io/mcis6273_f17_datamining/syllabus):

  * [Pandas Cookbook Github Repo](https://github.com/jvns/pandas-cookbook)
  * [Pandas Cheatsheet @ DataCamp.com](https://s3.amazonaws.com/assets.datacamp.com/blog_assets/PandasPythonForDataScience.pdf)

&#167;  In the readings and lecture, we talked about distributions of data.  Please produce the frequency distribution (histogram) for the following:  (1) rural income, (2) car ownership.  You will most certainly need to use the [`pandas.Series.plot.hist()`](http://pandas.pydata.org/pandas-docs/stable/visualization.html#visualization-hist) method.


&#167;  Produce the scatter plot for income vs age -- it doesn't matter which
is on the $x$ and $y$ axis.  You will benefit from the [`DataFrame.plot.scatter()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.plot.scatter.html#pandas.DataFrame.plot.scatter) method.



### (40%) Explore concepts in statistical inference over real data within Scikit-Learn 

**For questions 1-3**: Just answer the question over the entire data
as requested.

**For questions 4 and 5**: Consider this scenario ...

> The bank is looking for candidates who might be good for sending
mortgage loan information to -- that is they are looking for people
who are **not homeowners** (i.e. don't own a home), but who have
the income and other characteristics of potentially good
borrowers.

> Your task is to look at all the data for people who have no mortgage
and build the case for the profile of the borrower (from this dataset)
that they should start calling, sending mail and advertising to.

> Imagine you are a real-world analyst and will need to do the following:

* build the dataset that contains non-homeowners (i.e. no mortgage)
* report on the characteristics of non-homeowners based on what is being asked

In [1]:
import pandas as pd
df = pd.read_csv("./data/bank-data.csv")

&#167;  What is the median age of a `UNMARRIED`, `FEMALE`, `HOME OWNER` in the `SUBURBAN` region?

In [2]:
df.head()

id  age     sex      region   income married  children  car save_act  \
0  ID12101   48  FEMALE  INNER_CITY  17546.0      NO         1   NO       NO   
1  ID12102   40    MALE        TOWN  30085.1     YES         3  YES       NO   
2  ID12103   51  FEMALE  INNER_CITY  16575.4     YES         0  YES      YES   
3  ID12104   23  FEMALE        TOWN  20375.4     YES         3   NO       NO   
4  ID12105   57  FEMALE       RURAL  50576.3     YES         0   NO      YES   

  current_act mortgage  pep  
0          NO       NO  YES  
1         YES      YES   NO  
2         YES       NO   NO  
3         YES       NO   NO  
4          NO       NO   NO

In [3]:
df[(df.married=='NO') & (df.sex=='FEMALE') & (df.mortgage=='YES') & (df.region=='SUBURBAN')].age.describe()['50%']

52.0

&#167;  Given this bank datset what is the joint probability $\Pr( \mathrm{sex} = MALE \wedge \mathrm{income} \ge 50,000 )$?  How does this compare with $\Pr( \mathrm{sex} = FEMALE \wedge \mathrm{income} \ge 50,000 )$?

Recall: 
$$\Pr( A \wedge B ) = \Pr(A\big|B)  \Pr(B)$$

or 

$$\Pr(A \wedge B ) = \Pr(A) \Pr(B)$$, when $A$ and $B$ are independent. 

In [4]:
pr_male_incgt50k = \
    df[(df.sex=='MALE')&(df.income>=50000)].shape[0] / df[df.income>=50000].shape[0] 

In [5]:
pr_incgt50k = df[df.income>=50000].shape[0] / df.shape[0]

In [6]:
pr_joint = pr_male_incgt50k * pr_incgt50k
pr_joint

0.03

&#167;  What is  conditional probability $\Pr( car = \mathrm{TRUE} \big| sex = \mathrm{MALE} \wedge income \ge \mathrm{50,000} )$?

Since we have already computed the condition $sex=MALE \wedge income \ge 50000$, we need only find the number of these which have cars:

In [7]:
df[(df.car=='YES') & (df.sex=='MALE') & (df.income>=50000)]

id  age   sex      region   income married  children  car save_act  \
42   ID12143   67  MALE        TOWN  55716.5      NO         2  YES      YES   
206  ID12307   63  MALE  INNER_CITY  59409.1      NO         0  YES      YES   
226  ID12327   67  MALE  INNER_CITY  50474.6     YES         2  YES      YES   
236  ID12337   65  MALE        TOWN  52255.9      NO         2  YES      YES   
282  ID12383   64  MALE  INNER_CITY  54314.5     YES         1  YES      YES   
306  ID12407   63  MALE  INNER_CITY  52117.3      NO         2  YES      YES   
342  ID12443   64  MALE  INNER_CITY  53104.3     YES         0  YES      YES   
580  ID12681   63  MALE    SUBURBAN  51879.3     YES         2  YES      YES   

    current_act mortgage  pep  
42           NO       NO  YES  
206         YES       NO  YES  
226         YES       NO  YES  
236         YES       NO  YES  
282          NO       NO  YES  
306         YES       NO  YES  
342         YES       NO   NO  
580          NO      YES  YES

In [9]:
df[(df.car=='YES') & (df.sex=='MALE') & (df.income>=50000)].count()[0] / \
    df[(df.sex=='MALE') & (df.income>=50000)].count()[0]

0.44444444444444442

&#167;  If the lending requirements were a savings account and income greater
than $45,000, what about this data might make it difficult to justify
any campaign at all?  (**HINT:** most mortgages are 30-years in duration) Provide **concrete evidence** for your claims; you may do this in a variety of ways _including describing the statistic that leads you to your claim_.

In [10]:
df[(df.save_act=='YES') & (df.income>=45000)].describe()

age        income   children
count  74.000000     74.000000  74.000000
mean   61.189189  52170.395946   1.135135
std     4.898073   4678.395935   1.051083
min    47.000000  45031.900000   0.000000
25%    58.000000  48732.850000   0.000000
50%    63.000000  51291.800000   1.000000
75%    64.000000  56263.000000   2.000000
max    67.000000  63130.100000   3.000000

There are 74 people in this group.  The median age however is 61.19!  The min is 47!  Even if you consider the youngest of this group, they would be in their late 70s (47 + 30-year mortgage), making them well into retirement by the time the home was even paid off!  Thus, this is not a good group to consider.

&#167;  If the savings account requirement was ignored and the income
lowered to a minimum of $30,000, to whom (`MALE` or `FEMALE`) and in
which region (`RURAL`, `TOWN`, `INNER_CITY`, `SUBURBAN`) would
the bank likely have the best success?

There are several ways to look at this.  First, we'll just find the people under 47 and sort by region and sex, then take a look at their statistics.

In [11]:
df[(df.income>=30000)&(df.mortgage=='NO')&(df.age<47)].sort_values(by=['sex','region'])

id  age     sex      region   income married  children  car  \
222  ID12323   42  FEMALE  INNER_CITY  39308.7     YES         1   NO   
291  ID12392   44  FEMALE  INNER_CITY  39253.6      NO         0  YES   
371  ID12472   44  FEMALE  INNER_CITY  38453.7      NO         2   NO   
381  ID12482   46  FEMALE  INNER_CITY  32583.5     YES         2  YES   
538  ID12639   43  FEMALE  INNER_CITY  38784.0     YES         0   NO   
123  ID12224   35  FEMALE       RURAL  33028.3      NO         1   NO   
425  ID12526   44  FEMALE       RURAL  30488.7     YES         0   NO   
442  ID12543   43  FEMALE       RURAL  36281.0     YES         0  YES   
449  ID12550   38  FEMALE       RURAL  34061.4      NO         0  YES   
558  ID12659   38  FEMALE    SUBURBAN  31290.6     YES         0   NO   
233  ID12334   32  FEMALE        TOWN  30404.3     YES         0  YES   
383  ID12484   40  FEMALE        TOWN  34836.8     YES         1  YES   
447  ID12548   35  FEMALE        TOWN  30799.5     YES         2   NO   
587  ID12688   43  FEMALE        TOWN  31273.8      NO         1  YES   
85   ID12186   36    MALE  INNER_CITY  31683.1     YES         1  YES   
514  ID12615   37    MALE  INNER_CITY  33886.4      NO         0  YES   
552  ID12653   45    MALE  INNER_CITY  41107.2     YES         2  YES   
571  ID12672   40    MALE  INNER_CITY  36166.2     YES         0   NO   
445  ID12546   40    MALE       RURAL  31864.8     YES         0  YES   
541  ID12642   39    MALE       RURAL  37389.0     YES         2   NO   
34   ID12135   43    MALE    SUBURBAN  37521.9      NO         0   NO   
105  ID12206   43    MALE        TOWN  32395.5     YES         3  YES   
110  ID12211   36    MALE        TOWN  37330.5      NO         2   NO   
113  ID12214   36    MALE        TOWN  33630.6      NO         2  YES   
151  ID12252   43    MALE        TOWN  36432.8      NO         2  YES   
176  ID12277   40    MALE        TOWN  37558.5     YES         0  YES   
218  ID12319   39    MALE        TOWN  31693.5      NO         0  YES   
310  ID12411   46    MALE        TOWN  30658.7     YES         0   NO   
479  ID12580   46    MALE        TOWN  43395.5      NO         1  YES   

    save_act current_act mortgage  pep  
222      YES         YES       NO  YES  
291      YES          NO       NO  YES  
371      YES         YES       NO  YES  
381      YES         YES       NO   NO  
538      YES         YES       NO   NO  
123      YES         YES       NO  YES  
425      YES         YES       NO   NO  
442      YES         YES       NO   NO  
449      YES         YES       NO  YES  
558      YES         YES       NO   NO  
233      YES         YES       NO   NO  
383      YES         YES       NO  YES  
447       NO         YES       NO  YES  
587      YES          NO       NO  YES  
85       YES         YES       NO  YES  
514      YES          NO       NO  YES  
552      YES         YES       NO   NO  
571       NO          NO       NO   NO  
445      YES         YES       NO   NO  
541      YES         YES       NO  YES  
34       YES         YES       NO  YES  
105      YES         YES       NO   NO  
110      YES         YES       NO  YES  
113      YES         YES       NO  YES  
151       NO         YES       NO  YES  
176      YES         YES       NO   NO  
218      YES          NO       NO  YES  
310      YES         YES       NO   NO  
479      YES         YES       NO  YES

The largest, youngest group are `MALES` in `TOWN`.  They have the income and the risk profile (e.g. savings account).  However the median age is an issue:

In [12]:
df[(df.income>=30000)&(df.mortgage=='NO')&(df.age<47)&(df.sex=='MALE')&(df.region=='TOWN')].describe()['age']['50%']

41.5

They'd be in their 70s by the time the mortgage was paid off.  Perhaps we should try to see what's going on with those just a but younger ...

We could also go to the below 35 age since many people ideally would like to have their home paid off before US retirement age.

In [13]:
df[(df.income>=30000)&(df.mortgage=='NO')&(df.age<35)].sort_values(by=['sex','region'])

id  age     sex region   income married  children  car save_act  \
233  ID12334   32  FEMALE   TOWN  30404.3     YES         0  YES      YES   

    current_act mortgage pep  
233         YES       NO  NO

In this group we only have `FEMALE` in `TOWN` but only one, so a campaign may not be worth the time.

### (40%) Work with data to understand distance metrics in Scikit-Learn and the impact various metrics have on the outcomes 

A common thing to understand in a dataset is to determine some
number of _nearest neighbors_ to a data point.  We will see this come
back when we get to clustering.  For now, let's explore the
[NearestNeighbor](http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html#sklearn.neighbors.NearestNeighbors)
implementation in Scikit-Learn.

You will most certainly need to convert the categorical non-numeric data
to binary features, and you may be served well by reading about
[preprocessing data in Scikit-Learn](http://scikit-learn.org/stable/modules/preprocessing.html#preprocessing) and also taking
a closer look at [`sklearn.preprocessing.LabelBinarizer`](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelBinarizer.html#sklearn.preprocessing.LabelBinarizer) and [`sklearn.preprocessing.OneHotEncoder`](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html#sklearn.preprocessing.OneHotEncoder).  You might also find
[this resource](https://courses.edx.org/c4x/BerkeleyX/CS190.1x/asset/CS190.1x_week4b.pdf)
of value if you read only the first few slides on binarization.

&#167;  What are the 5 nearest neighbors (the indices) to index #7, #21, #40
and #94?  Your output will be a list of the tuples
of the 5 closest indices and their values (e.g. `[(5, 4.56356), (19, 8.83452), (233, 12.23486), ...]`).  Use the _minkowski
(default) distance_ first.


&#167;  Use the _cosine similarity metric_ and _euclidean distance metric_
(e.g. invoke `NearestNeighbor(..., metric='cosine')` and `NearestNeighbor(..., metric='euclidean')`).
Produce the same lists for the same indices as in #1.  What are
the differences in the nearest neighbor lists?  Which seem to be
the most similar? Provide your thoughts on why there are differences?

In [14]:
df = df.drop('id', 1)
df_binarized = pd.get_dummies(df)

In [15]:

def compute_nn_report(nn_list=[7, 21, 40, 90], metric=None, k=5):
    from sklearn.neighbors import NearestNeighbors
    
    if metric:
        nn = NearestNeighbors(k, metric=metric, algorithm='brute')        
    else:
        nn = NearestNeighbors(k)
    
    nn_matrix = nn.fit(df_binarized).kneighbors(df_binarized)

    for i in nn_list:
        print("The nearest neighbors to {} are: {} \n\tevidence: {}" \
              .format(i, nn_matrix[1][i], tuple(zip(nn_matrix[1][i], nn_matrix[0][i]))))

In [16]:
compute_nn_report(metric='minkowski')

The nearest neighbors to 7 are: [  7  56 164 227 565] 
	evidence: ((7, 0.0), (56, 31.09356846588663), (164, 46.354719284992584), (227, 58.485553771361175), (565, 83.827203221154832))
The nearest neighbors to 21 are: [ 21 436 413   2 483] 
	evidence: ((21, 0.0), (436, 23.618636709881851), (413, 32.306191357038912), (2, 78.170390813623925), (483, 97.371710471058307))
The nearest neighbors to 40 are: [ 40 100 168 335 599] 
	evidence: ((40, 0.0), (100, 66.345233437964566), (168, 69.172899317476165), (335, 86.828624312025909), (599, 104.01807535169671))
The nearest neighbors to 90 are: [ 90 377  60 559 565] 
	evidence: ((90, 0.0), (377, 10.04987562112089), (60, 51.521257748020972), (559, 53.23729895401285), (565, 53.643359328199281))


In [17]:
compute_nn_report(metric='euclidean')

The nearest neighbors to 7 are: [  7  56 164 227 565] 
	evidence: ((7, 0.0), (56, 31.09356846588663), (164, 46.354719284992584), (227, 58.485553771361175), (565, 83.827203221154832))
The nearest neighbors to 21 are: [ 21 436 413   2 483] 
	evidence: ((21, 0.0), (436, 23.618636709881851), (413, 32.306191357038912), (2, 78.170390813623925), (483, 97.371710471058307))
The nearest neighbors to 40 are: [ 40 100 168 335 599] 
	evidence: ((40, 0.0), (100, 66.345233437964566), (168, 69.172899317476165), (335, 86.828624312025909), (599, 104.01807535169671))
The nearest neighbors to 90 are: [ 90 377  60 559 565] 
	evidence: ((90, 0.0), (377, 10.04987562112089), (60, 51.521257748020972), (559, 53.23729895401285), (565, 53.643359328199281))


In [18]:
compute_nn_report(metric='cosine')

The nearest neighbors to 7 are: [  7  38 592 524 227] 
	evidence: ((7, 2.2204460492503131e-16), (38, 2.9460347494847383e-09), (592, 3.2576520370142248e-09), (524, 7.5093011808746724e-09), (227, 7.6371569068811596e-09))
The nearest neighbors to 21 are: [ 21 202 128 132 314] 
	evidence: ((21, 0.0), (202, 1.5016694221436921e-08), (128, 1.6308320693880773e-08), (132, 1.7706350607937793e-08), (314, 1.8218509589829068e-08))
The nearest neighbors to 40 are: [ 40 533 100 141 387] 
	evidence: ((40, 0.0), (533, 3.1807753098078706e-09), (100, 4.020875410404301e-09), (141, 6.7570852246845448e-09), (387, 7.3089312380147931e-09))
The nearest neighbors to 90 are: [ 90 185 194 517 599] 
	evidence: ((90, 2.2204460492503131e-16), (185, 2.9281650437695816e-09), (194, 3.3899721918473347e-09), (517, 4.4379550034889803e-09), (599, 4.5540711202107786e-09))
